src/main/src/main/java/org/geoserver/catalog/impl/ResolvingProxyResolver.java

```diff
@@ -0,0 +1,284 @@
+/* (c) 2014 Open Source Geospatial Foundation - all rights reserved
+ * (c) 2001 - 2013 OpenPlans
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.catalog.impl;
+
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.geoserver.catalog.Catalog;
+import org.geoserver.catalog.CatalogInfo;
+import org.geoserver.catalog.LayerGroupInfo;
+import org.geoserver.catalog.LayerInfo;
+import org.geoserver.catalog.MapInfo;
+import org.geoserver.catalog.NamespaceInfo;
+import org.geoserver.catalog.PublishedInfo;
+import org.geoserver.catalog.ResourceInfo;
+import org.geoserver.catalog.StoreInfo;
+import org.geoserver.catalog.StyleInfo;
+import org.geoserver.catalog.WorkspaceInfo;
+import org.geotools.util.logging.Logging;
+
+/**
+ * Utility class for resolving proxied catalog objects.
+ *
+ * <p>This class provides methods to replace proxy references in catalog objects with actual instances from the catalog.
+ * It recursively resolves references within complex catalog objects such as layers, layer groups, and resources to
+ * ensure all nested objects are properly instantiated.
+ *
+ * <p>The resolver handles special cases during catalog loading where certain references may not yet be available,
+ * maintaining proxies when resolution isn't possible. It also unwraps modification proxies to ensure direct object
+ * references are used.
+ *
+ * <p>Use this class when:
+ *
+ * <ul>
+ *   <li>Loading objects from persistent storage
+ *   <li>Preparing objects for serialization
+ *   <li>Ensuring consistent object references throughout the catalog
+ * </ul>
+ *
+ * <p>Each catalog object type has a specialized resolution method to handle its specific reference structure.
+ */
+public class ResolvingProxyResolver {
+
+    private static final Logger LOGGER = Logging.getLogger(ResolvingProxyResolver.class);
+
+    /**
+     * Private constructor to prevent instantiation.
+     *
+     * <p>This is a utility class that provides only static methods and should not be instantiated.
+     */
+    private ResolvingProxyResolver() {
+        // private constructor, this is an utility class
+    }
+
+    /**
+     * Unwraps a modification proxy to access the underlying object.
+     *
+     * <p>This helper method delegates to {@link ModificationProxy#unwrap(Object)} to ensure we're working with the
+     * actual object instances rather than their proxy wrappers.
+     *
+     * @param <T> the type of the object being unwrapped
+     * @param obj the potentially proxied object
+     * @return the unwrapped object, or the original object if not a proxy
+     */
+    private static <T> T unwrap(T obj) {
+        return ModificationProxy.unwrap(obj);
```

Comment:
not sure I get the value of having this method here rather than using static method ModificationProxy.unwrap(obj) directly?

Reply:
It's only a refactoring of what was in AbstractCatalogFacade, no judgements.